src/main/java/io/gravitee/policy/groovy/GroovyPolicy.java

```diff
@@ -15,250 +15,137 @@
+@Slf4j
+public class GroovyPolicy extends GroovyPolicyV3 implements Policy {
 
-    private final GroovyPolicyConfiguration groovyPolicyConfiguration;
+    private static final ExecutionFailure DEFAULT_EXECUTION_FAILURE = new ExecutionFailure(INTERNAL_SERVER_ERROR_500)
+        .key("GROOVY_EXECUTION_FAILURE")
+        .message("Groovy execution failure");
 
-    private static final String REQUEST_VARIABLE_NAME = "request";
-    private static final String RESPONSE_VARIABLE_NAME = "response";
-    private static final String CONTEXT_VARIABLE_NAME = "context";
-    private static final String RESULT_VARIABLE_NAME = "result";
+    private final Flowable<String> scriptFlowable;
 
-    private static final SecuredGroovyShell GROOVY_SHELL = new SecuredGroovyShell();
+    public GroovyPolicy(GroovyPolicyConfiguration configuration) {
+        super(configuration);
+        scriptFlowable = Flowable.fromIterable(configuration.getScripts());
+    }
 
-    public GroovyPolicy(GroovyPolicyConfiguration groovyPolicyConfiguration) {
-        this.groovyPolicyConfiguration = groovyPolicyConfiguration;
+    @Override
+    public String id() {
+        return "policy-groovy";
     }
 
-    @OnRequest
-    public void onRequest(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        executeScript(request, response, executionContext, policyChain, groovyPolicyConfiguration.getOnRequestScript());
+    @Override
+    public Completable onRequest(HttpExecutionContext ctx) {
+        return ctx
+            .request()
+            .onBody(bodyBuffer ->
+                bodyBuffer
+                    .defaultIfEmpty(Buffer.buffer())
+                    .flatMapMaybe(buffer -> onHttp(ctx, buffer, ctx.request().headers(), GroovyBindings.bindRequest(ctx, buffer)))
+            );
     }
 
-    @OnResponse
-    public void onResponse(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        executeScript(request, response, executionContext, policyChain, groovyPolicyConfiguration.getOnResponseScript());
+    @Override
+    public Completable onResponse(HttpExecutionContext ctx) {
+        return ctx
+            .response()
+            .onBody(bodyBuffer ->
+                bodyBuffer
+                    .defaultIfEmpty(Buffer.buffer())
+                    .flatMapMaybe(buffer -> onHttp(ctx, buffer, ctx.response().headers(), GroovyBindings.bindResponse(ctx, buffer)))
+            );
     }
 
-    @OnResponseContent
-    public ReadWriteStream onResponseContent(
-        Request request,
-        Response response,
-        ExecutionContext executionContext,
-        PolicyChain policyChain
-    ) {
-        String script = groovyPolicyConfiguration.getOnResponseContentScript();
+    private Maybe<Buffer> onHttp(HttpExecutionContext ctx, Buffer bodyBuffer, HttpHeaders headers, Binding binding) {
+        return scriptFlowable
```

Comment:
Does it execute all the scripts? I also see that only the last is retained but I have difficulty understanding.